#533 DeltaFunction

zunyibrt opened the pull request:
Implementation for a DeltaFunction SBProfile

rmandelb commented:
FYI, @dkirkby - 3ish years after your request, we finally have a DeltaFunction base class...!  (still to be reviewed)

rmjarvis commented on tests/test_base.py:
```diff
@@ -85,7 +248,7 @@ def test_gaussian():
             err_msg="Using GSObject Gaussian disagrees with expected result")
     np.testing.assert_almost_equal(
-            myImg.array.sum(dtype=float) *dx**2, myImg.added_flux, 5,
+            myImg.array.sum() *dx**2, myImg.added_flux, 5,
```
I don't think you meant to change this.  It looks like you had some kind of merge error and ended up committing back to an earlier version of a file from here down.  You should copy back in the master version of this file for the parts that aren't new for DeltaFunction.

rmjarvis commented on tests/test_base.py:
```diff
@@ -69,6 +69,169 @@
+    np.testing.assert_equal(delta.kValue(cen), (1+0j) * test_flux)
+    np.testing.assert_equal(delta.kValue(offcen), (1+0j) * test_flux)
+    import math
+    np.testing.assert_almost_equal(delta.xValue(cen), test_flux)
```
Again, this isn't right.

rmjarvis commented on tests/test_base.py:
```diff
@@ -69,6 +69,169 @@
+    np.testing.assert_almost_equal(delta.xValue(cen), test_flux)
+    np.testing.assert_almost_equal(delta.xValue(offcen), 0)
+    np.testing.assert_array_less(1e100,delta.maxK())
+    np.testing.assert_array_less(1e100,delta.stepK())
```
Rather than `assert_array_less` here, you can just do `assert delta.stepK() < 1.e100`.  Not everything has to go through the numpy testing module.  :)

rmjarvis commented on tests/test_base.py:
```diff
@@ -69,6 +69,169 @@
+    np.testing.assert_almost_equal(
+        obj2.getFlux(), test_flux * 2., decimal=param_decimal,
+        err_msg="Flux param inconsistent after obj.withFlux(flux).")
+        
```
Some other things that might be worth testing:
* I don't see any tests of the shooting code.  Should add a call to `do_shoot` for a DeltaFunction convolved with something simple.  
* Maybe convolve more than one DeltaFunction together.  fluxes should multiply.
* Rather than test that expand, shear, etc. give errors, I think they should be no ops, so check that a sheared DeltaFunction convolved with something is equivalent to not shearing it.  (shearing just the DeltaFunction that is, not the other component.)
* There is a function called `check_basic` that does a number of sanity checks.  It's possible that DeltaFunction may fail for this, since it might make assumptions about the way objects work that don't apply here.  But at least try it out and if necessary add a kwarg to skip the basic tests that don't really make sense for DeltaFunction.
* Maybe try a real-space convolution, especially if you try the thing I suggested above about making xValue near zero not inf, but rather flux / eps^2.  You can test this with `Convolve(delta, gauss, real_space=True)`.
* I don't think you anywhere construct `DeltaFunction()` with no arguments.  Should do that and make sure it has flux=1.  Also make sure it pickles correctly, which will cover that branch in `__str__`.

rmandelb commented:
@zunyibrt - do you have updates in response to the comments so far?

zunyibrt commented:
Hi, I made the changes based on comments and added the suggested test cases. I also followed the suggestion and tried the real space convolution but it didn't seem to be doing it correctly, so I removed it.

zunyibrt commented:
@rmjarvis , the Travis CI build check seems to be failing for python 3.6. I tried reading the logs but it seems that it timed out so I can't seem to figure out the issue.

rmjarvis commented:
Thanks Brent.  I made a few minor edits.  But I think this is ready to go now.  Let's plan to merge Friday to give anyone else who wants to take a look time to do so.

rmandelb commented:
Thanks.  I will plan to take a look during this week.

rmandelb commented on include/galsim/SBDeltaFunctionImpl.h:
```diff
@@ -0,0 +1,79 @@
+        /**
+         * @brief Shoot photons through this SBDeltaFunction.
+         *
+         * SBDeltaFunction shoots photons by analytic transformation of the unit disk.  Slightly
```
Is this actually true or was this copy/pasted from SBGaussian?

zunyibrt replied:
copy/pasted. Removed the comment.

rmandelb commented on src/SBDeltaFunction.cpp:
```diff
@@ -0,0 +1,104 @@
+#include "SBDeltaFunction.h"
+#include "SBDeltaFunctionImpl.h"
+
+// Define this variable to find azimuth (and sometimes radius within a unit disc) of 2d photons by
```
I think this entire commented paragraph doesn't belong here (copy/paste issue).

zunyibrt replied:
Yup. Removed.

rmandelb commented on tests/test_config_gsobject.py:
```diff
@@ -78,6 +79,11 @@ def test_gaussian():
     gal5b = gal5b.rotate(-34 * galsim.degrees).lens(-0.15, 0.2, 0.93)
     gsobject_compare(gal5a, gal5b)
 
+    # DeltaFunction is functionally equivalent to an extremely narrow Gaussian.
```
This is true, but shouldn't we also compare with an actual DeltaFunction in addition to comparing with an extremely narrow Gaussian?

rmjarvis replied:
Sure.  Done.

rmandelb replied:
Thanks.

rmandelb commented:
Also, @rmjarvis - I noticed that on this branch I'm getting a warning that I'm not getting on master when I try to run unit tests:
```
/Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/site-packages/galsim/position.py:22: RuntimeWarning: to-Python converter for galsim::Silicon already registered; second conversion method ignored.
  from . import _galsim
/Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/site-packages/galsim/position.py:22: RuntimeWarning: to-Python converter for boost::shared_ptr<galsim::Silicon> already registered; second conversion method ignored.
  from . import _galsim
```
Seems to have nothing to do with DeltaFunction but I'm not sure why it could be happening here and not on master.

rmjarvis commented:
>  I noticed that on this branch I'm getting a warning that I'm not getting on master when I try to run unit tests:

I also don't know why that started showing up.  But I fixed it on '#865' here: b4181f83de54417b6c19e6b4b17ccdbb7e221c4c

rmandelb commented:
Thanks.

rmjarvis commented on include/galsim/SBDeltaFunctionImpl.h:
```diff
@@ -55,12 +55,7 @@ namespace galsim {
-         * (sqrt,etc.)
-         *
          * @param[in] N Total number of photons to produce.
-         * @param[in] ud UniformDeviate that will be used to draw photons from distribution.
```
I don't think you meant to remove this line.

zunyibrt replied:
Yes, did not mean to remove it. Added it back.

rmjarvis commented:
@rmandelb I think this is just waiting for your final approval.

rmandelb commented:
Sorry - done :)
